packages/runtime/lib/dependencies.js

```diff
@@ -0,0 +1,126 @@
+'use strict'
+const { readFile } = require('node:fs/promises')
+const { resolve } = require('node:path')
+const { closest } = require('fastest-levenshtein')
+const Topo = require('@hapi/topo')
+const ConfigManager = require('@platformatic/config')
+const errors = require('./errors')
+
+function missingDependencyErrorMessage (clientName, service, config) {
+  const closestName = closest(clientName, [...config.serviceMap.keys()])
+  let errorMsg = `service '${service.id}' has unknown dependency: '${clientName}'.`
+  if (closestName) {
+    errorMsg += ` Did you mean '${closestName}'?`
+  }
+  return errorMsg
+}
+
+async function getClientId (service, client) {
+  try {
+    const clientPackageJsonPath = resolve(service.path, client.path, 'package.json')
+    const clientPackageJSON = JSON.parse(await readFile(clientPackageJsonPath, 'utf8'))
+    return clientPackageJSON.name ?? ''
+  } catch (err) {
+    if (client.url === undefined || client.name === undefined) {
+      throw err
+    }
+  }
+}
+
+async function parseConfigUrl (configManager, urlString) {
+  if (!urlString) {
+    return { url: null, envVar: null }
+  }
+
+  try {
+    const url = await configManager.replaceEnv(urlString)
+    return { url, envVar: null }
+  } catch (err) {
+    // The MissingValueError is an error coming from pupa
+    // https://github.com/sindresorhus/pupa#missingvalueerror
+    // All other errors are simply re-thrown.
+    if (err.name !== 'MissingValueError' || urlString !== `{${err.key}}`) {
+      throw err
+    }
+
+    return { url: null, envVar: err.key }
+  }
+}
+
+async function addServiceDependency (configManager, config, service, dependencyId, urlString) {
+  let { url, envVar } = await parseConfigUrl(configManager, urlString)
+  if (url !== null) {
+    service.dependencies.push({ id: dependencyId, url, local: false })
+    return
+  }
+
+  const depService = config.serviceMap.get(dependencyId)
+  if (depService === undefined) {
+    throw new errors.MissingDependencyError(
+      missingDependencyErrorMessage(dependencyId, service, config)
+    )
+  }
+
+  url = `http://${dependencyId}.plt.local`
+
+  if (envVar !== null) {
+    service.localServiceEnvVars.set(envVar, url)
+  }
+
+  depService.dependents.push(service.id)
+  service.dependencies.push({ id: dependencyId, url, local: true })
+}
+
+async function parseClientsAndComposer (config) {
```

Comment:
Can we move these to service and composer?

Reply:
+1 for this one. And I would create an interface. Somethinlg like `service.getDependencies`, so each stackable/template can parse it's own dependencies from different places.

Reply:
It is done now!